### PR TITLE
Add information about nest_vars.sh and update Linux installation instructions

### LINKED ISSUE #1222: Add `LD_LIBRARY_PATH` to guide for "writing an extension module"
To build an extension module for NEST, I followed the guide on the NEST Developer Space: https://nest.github.io/nest-simulator/extension_modules

However, after following steps 1-5 for installing the example module, I couldn't load it, but received this error: 
```Jun 06 09:23:59 Install [Error]: DynamicModuleManagementError
    Module 'mymodule' could not be opened.
    The dynamic loader returned the following error: 'file not found'.
```

I could easily fix this by adding the NEST lib folder to my `LD_LIBRARY_PATH`, like so:
`export LD_LIBRARY_PATH=~/opt/nest/ins/lib/nest`

Now, perhaps it would be a good idea to add this simple step to the documentation?

### PR DESCRIPTION
**Pull request by sarakonradi**
This PR adds information on setting the module loading path correctly.

More specifically, it:

- Adds a ``Setting the module loading path`` section below ``Installation from source``
- Removes the duplicate information in the ``IMPORTANT!`` admonition
- Turns these instructions into numbered steps for a better user experience

It fixes #1222.

### CODE REVIEW

**Comment by sarakonradi** [comment]
@jougs Does this address your concern? Or do you suggest adding a section about ``nest_vars.sh`` in general, explaining all the paths that it entails?

File: doc/installation/linux_install.rst [hunk 1]
```diff
@@ -71,15 +71,22 @@ The following are the basic steps to compile and install NEST from source code:
...
+In order to correctly set the module loading path or build an extension module for NEST:
+
+1. Source ``nest_vars.sh``:
 
   .. code-block:: sh
 
-       source </path/to/nest_install_dir>/bin/nest_vars.sh
+      source </path/to/nest_install_dir>/bin/nest_vars.sh
+
+2. Set the ``NEST_MODULE_PATH``.
```
Comment (terhorstd) [inline doc/installation/linux_install.rst:88]: I would include a line how to actually do it:
```suggestion
2. Set the ``NEST_MODULE_PATH``.

  .. code-block:: sh
      export NEST_MODULE_PATH=</path/to/your/modules/>
      
```

**Comment by clinssen** [comment]
@sarakonradi: could you set this PR to draft mode?

Waiting for my PR on @sarakonradi's fork to be merged (see https://github.com/sarakonradi/nest-simulator/pull/2/files) following updates in #1513.

Please feel free to leave any comments/suggestions there!

**Comment by sarakonradi** [comment]
@clinssen Sure! Just set it to draft mode.

It seems that your PR still includes the old information. For example, ``Python`` has been changed to ``Python 3.X`` and ``NEST requires at least version v2.8.12 of cmake`` has been updated to ``NEST requires CMake 3.12 or higher``: https://github.com/nest/nest-simulator/blob/master/doc/installation/linux_install.rst

Could you perhaps merge master again and then let me know?

**Comment by sarakonradi** [comment]
@clinssen Thanks for rebasing. Just merged your PR and will keep an eye on #1513.

File: doc/installation/linux_install.rst [hunk 2]
```diff
@@ -3,11 +3,29 @@ Ubuntu/Debian Installation
...
 .. _standard:
 
-Installation from source
---------------------------
+Dependencies
+------------
+
+To build NEST, you need a recent version of `CMake <https://cmake.org>`_ and `libtool <https://www.gnu.org/software/libtool/libtool.html>`_; the latter should be available for most systems and is probably already installed.
+
+.. note:: NEST requires CMake 3.12 or higher, but we recommend version 3.16.X. You can type ``cmake --version`` on the commandline to check your current version.
```
Comment (clinssen) [inline doc/installation/linux_install.rst:11]:
```suggestion
.. note:: NEST requires CMake 3.12 or higher, but we recommend version 3.16.X. You can type ``cmake --version`` on the commandline to check your current version.

```

File: doc/installation/linux_install.rst [hunk 3]
```diff
@@ -129,3 +116,40 @@ If you want to run the ``nest`` executable or use the ``nest`` Python module wit
...
+   * - ``NEST_DOC_DIR``
+     - NEST built-in online help finds help files ``$NEST_DOC_DIR/help``
+   * - ``NEST_PYTHON_PREFIX``
+     - The path where the PyNEST bindings are installed.
+   * - ``PYTHONPATH``
+     - Search path for non-standard Python module locations. Will be prepended to or created to if it does not exist.
+   * - ``PATH``
+     - Search path for binaries. Will be prepended to or created to if it does not exist.
+   * - ``LD_LIBRARY_PATH``
+     - Search path for shared objects (*.so files). Note: called ``DYLD_LIBRARY_PATH`` on MacOS.  Will be prepended to or created to if it does not exist.
```
Comment (clinssen) [inline doc/installation/linux_install.rst:153]:
```suggestion
     - Search path for shared objects (*.so files). Note: called ``DYLD_LIBRARY_PATH`` on MacOS.  Will be prepended to or created if it does not exist.
```

File: doc/installation/linux_install.rst [hunk 4]
```diff
@@ -129,3 +116,40 @@ If you want to run the ``nest`` executable or use the ``nest`` Python module wit
...
+   * - ``NEST_INSTALL_DIR``
+     - NEST installation directory. Contains ``bin``, ``lib``, etc.
+   * - ``NEST_DATA_DIR``	
+     - NEST finds standard *.sli files ``$NEST_DATA_DIR/sli``
+   * - ``NEST_DOC_DIR``
+     - NEST built-in online help finds help files ``$NEST_DOC_DIR/help``
+   * - ``NEST_PYTHON_PREFIX``
+     - The path where the PyNEST bindings are installed.
+   * - ``PYTHONPATH``
+     - Search path for non-standard Python module locations. Will be prepended to or created to if it does not exist.
```
Comment (clinssen) [inline doc/installation/linux_install.rst:149]:
```suggestion
     - Search path for non-standard Python module locations. Will be prepended to or created if it does not exist.
```

File: doc/installation/linux_install.rst [hunk 5]
```diff
@@ -129,3 +116,40 @@ If you want to run the ``nest`` executable or use the ``nest`` Python module wit
...
+   * - ``PATH``
+     - Search path for binaries. Will be prepended to or created to if it does not exist.
```
Comment (clinssen) [inline doc/installation/linux_install.rst:151]:
```suggestion
     - Search path for binaries. Will be prepended to or created if it does not exist.
```

File: doc/installation/linux_install.rst (hunk 1 above)
Comment (sarakonradi) [reply doc/installation/linux_install.rst]: @terhorstd This file got updated after the merge of sarakonradi/nest-simulator#2. Please check again!

**Comment by sarakonradi** [comment]
Just added the CMake installation link, as suggested by @gtrensch. 

@terhorstd @clinssen @jougs Please feel free to have another look!

**Comment by clinssen** [comment]
Actually, it seems that Python 3.5 is already too old to be supported by NEST: `hl_api_connection_helpers.py` uses Python 3.6 f-strings feature (see https://github.com/nest/nest-simulator/blob/9ca796ef014058d01240bb977954e6b3e9d6b841/pynest/nest/lib/hl_api_connection_helpers.py#L75), which was introduced in Python 3.6.

**Comment by heplesser** [comment]
> Actually, it seems that Python 3.5 is already too old to be supported by NEST: `hl_api_connection_helpers.py` uses Python 3.6 f-strings feature (see
> 
> https://github.com/nest/nest-simulator/blob/9ca796ef014058d01240bb977954e6b3e9d6b841/pynest/nest/lib/hl_api_connection_helpers.py#L75
> 
> ), which was introduced in Python 3.6.

... which raises the question of which Python 3 version is the oldest we want to support. Discussion tomorrow!

**Comment by jougs** [comment]
@heplesser: We had that discussion already at the hackathon in November. Our conclusion is that we only support [the ones that did not reach their end of life](https://devguide.python.org/#status-of-python-branches). This means we should still support Python 3.5 until 2020-09-13.

**Comment by heplesser** [comment]
> @heplesser: We had that discussion already at the hackathon in November. Our conclusion is that we only support [the ones that did not reach their end of life](https://devguide.python.org/#status-of-python-branches). This means we should still support Python 3.5 until 2020-09-13.

@jougs Thanks for the reminder! Then we need to replace the f-strings.

**Comment by clinssen** [comment]
This should also affect the CI environment, since Travis is currently running with Python 3.6.7 on Linux and 3.7.5 on MacOS. @lekshmideepu: just a heads up!

**Comment by sarakonradi** [comment]
Just merged @clinssen's PR, which updates the package versions to Python 3. Please check!

**Comment by clinssen** [comment]
Reviewing this PR would be unethical as most of the content comes from me!

**Comment by sarakonradi** [comment]
@clinssen That is true! I seem to be unable to remove you as a reviewer entirely, but we can keep that in mind whenever we check the "Reviewers" list.

File: doc/installation/linux_install.rst [hunk 6]
```diff
@@ -69,55 +86,26 @@ The following are the basic steps to compile and install NEST from source code:
...
-
-.. admonition:: IMPORTANT!
-
- If your operating system does not find the ``nest`` executable or if Python does not find the ``nest`` module, your path variables may not be set correctly. This may also be the case if Python cannot load the ``nest`` module due to missing or incompatible libraries. In this case, please run:
-
-  .. code-block:: sh
+NEST should now be successfully installed on your system.
 
-       source </path/to/nest_install_dir>/bin/nest_vars.sh
+* Before using NEST, make sure that all the environment variables are set correctly. See the section `Environment variables`_ for details.
```
Comment (jougs) [inline doc/installation/linux_install.rst:91]:
```suggestion
* Before using NEST, make sure that all required environment variables are set correctly. See the section `Environment variables`_ for details.
```

Alternatively, the following could be a nice shortcut for people who already know their way around:

```suggestion
* Before using NEST, make sure that all required environment variables are set correctly. In short, this can be established by sourcing the shell script `nest_vars.sh`, which is installed into the path for binaries selected during the CMake run. See the section `Environment variables`_ for details.
```

File: doc/installation/linux_install.rst [hunk 7]
```diff
@@ -129,3 +117,40 @@ If you want to run the ``nest`` executable or use the ``nest`` Python module wit
 The script ``/install/path/bin/nest_vars.sh`` can be sourced in ``.bashrc`` and will set these paths for you. This also allows to switch between NEST installations in a convenient manner.
 
 
+Environment variables
+---------------------
+
+There are several environment variables that describe where components of the NEST installation can be found. In particular when installing to a custom directory, it is typically necessary to explicitly set these variables, so that your operating system can find the NEST binaries and libraries.
```
Comment (jougs) [inline doc/installation/linux_install.rst:123]:
```suggestion
A number of environment variables are used to specify where the components of a NEST installation are found. In particular when installing to a custom directory, it is typically necessary to explicitly set these variables, so that your operating system can find the NEST binaries, its libraries and custom extension modules.
```

File: doc/installation/linux_install.rst [hunk 8]
```diff
@@ -129,3 +117,40 @@ If you want to run the ``nest`` executable or use the ``nest`` Python module wit
...
+
+A shell script is provided in ``</install/path/>bin/nest_vars.sh`` to make setting the environment variables more convenient. Setting the environment variables in your active shell session requires sourcing the script:
```
Comment (jougs) [inline doc/installation/linux_install.rst:125]:
```suggestion
For your convenience, a shell script setting all required environment variables is provided in ``</install/path/>bin/nest_vars.sh``. Setting the environment variables in your active shell session requires sourcing the script:
```

File: doc/installation/linux_install.rst [hunk 9]
```diff
@@ -129,3 +117,40 @@ If you want to run the ``nest`` executable or use the ``nest`` Python module wit
...
+
+.. code-block:: sh
+
+   source </install/path/>bin/nest_vars.sh
+
+You may want to include this line in your ``.bashrc`` file, so that the environment variables are set automatically.
```
Comment (jougs) [inline doc/installation/linux_install.rst:131]:
```suggestion
You may want to include this line in your ``.bashrc`` file, so that the environment variables are set automatically whenever you open a new terminal.
```

File: doc/installation/linux_install.rst [hunk 10]
```diff
@@ -129,3 +117,40 @@ If you want to run the ``nest`` executable or use the ``nest`` Python module wit
...
+
+The following variables are set in ``nest_vars.sh``:
+
+.. list-table::
+   :header-rows: 1
+   :widths: 10 30
+
+   * - Path
```
Comment (jougs) [inline doc/installation/linux_install.rst:139]:
```suggestion
   * - Variable
```

File: doc/installation/linux_install.rst [hunk 11]
```diff
@@ -129,3 +117,40 @@ If you want to run the ``nest`` executable or use the ``nest`` Python module wit
...
+     - Description
+   * - ``NEST_INSTALL_DIR``
+     - NEST installation directory. Contains ``bin``, ``lib``, etc.
+   * - ``NEST_DATA_DIR``	
+     - NEST finds standard *.sli files ``$NEST_DATA_DIR/sli``
```
Comment (jougs) [inline doc/installation/linux_install.rst:144]:
```suggestion
     - NEST finds standard *.sli files in ``$NEST_DATA_DIR/sli``
```

File: doc/installation/linux_install.rst [hunk 12]
```diff
@@ -129,3 +117,40 @@ If you want to run the ``nest`` executable or use the ``nest`` Python module wit
...
+   * - ``NEST_DOC_DIR``
+     - NEST built-in online help finds help files ``$NEST_DOC_DIR/help``
```
Comment (jougs) [inline doc/installation/linux_install.rst:146]:
```suggestion
     - NEST built-in online help finds help files in ``$NEST_DOC_DIR/help``
```

File: doc/installation/linux_install.rst [hunk 13]
```diff
@@ -129,3 +117,40 @@ If you want to run the ``nest`` executable or use the ``nest`` Python module wit
...
+   * - ``NEST_PYTHON_PREFIX``
+     - The path where the PyNEST bindings are installed.
+   * - ``PYTHONPATH``
+     - Search path for non-standard Python module locations. Will be prepended to or created if it does not exist.
```
Comment (jougs) [inline doc/installation/linux_install.rst:150]:
```suggestion
     - Search path for non-standard Python module locations. Will be newly set or prepended to the already existing variable if it is already set.
```

File: doc/installation/linux_install.rst [hunk 14]
```diff
@@ -129,3 +117,40 @@ If you want to run the ``nest`` executable or use the ``nest`` Python module wit
...
+   * - ``PATH``
+     - Search path for binaries. Will be prepended to or created if it does not exist.
```
Comment (jougs) [inline doc/installation/linux_install.rst:152]:
```suggestion
     - Search path for binaries. Will be newly set or prepended to the already existing variable if it is already set.
```

File: doc/installation/linux_install.rst [hunk 15]
```diff
@@ -129,3 +117,40 @@ If you want to run the ``nest`` executable or use the ``nest`` Python module wit
...
+   * - ``LD_LIBRARY_PATH``
+     - Search path for shared objects (*.so files). Note: called ``DYLD_LIBRARY_PATH`` on MacOS.  Will be prepended to or created if it does not exist.
```
Comment (jougs) [inline doc/installation/linux_install.rst:154]:
```suggestion
     - Search path for shared objects (*.so files). Note: called ``DYLD_LIBRARY_PATH`` on MacOS. Will be newly set or prepended to the already existing variable if it is already set.
```

**Comment by sarakonradi** [comment]
> Very nice, thanks! I still have some suggestions for you to consider.

Thanks for your suggestions! Just applied them. I also fixed the merge conflict. ``doc/installation/linux_install.rst`` now includes the note on python bindings being enabled by default (#1507).

@jougs @terhorstd Please check!

**Comment by jougs** [comment]
@terhorstd are you also happy? If yes, please also click the button. Thanks!

File: doc/installation/linux_install.rst [hunk 16]
```diff
@@ -1,29 +1,50 @@
...
+==========================
 
 .. _standard:
 
-Installation from source
---------------------------
+Dependencies
+------------
+
+To build NEST, you need a recent version of `CMake <https://cmake.org/install>`_ and `libtool <https://www.gnu.org/software/libtool/libtool.html>`_; the latter should be available for most systems and is probably already installed.
```
Comment (terhorstd) [inline doc/installation/linux_install.rst:9]: We should have a central file where dependencies are collected, as we have for the docs (conda environment). This section could then have an additional sentence "For the full list of requirements check [here]()". We should do this in a separate PR though.

**Comment by sarakonradi** [comment]
@terhorstd I created a separate issue (#1612) to address your concern regarding the dependencies file.

File: doc/installation/linux_install.rst (hunk 16 above)
Comment (terhorstd) [reply doc/installation/linux_install.rst:9]: Thanks for adding #1612 for this, @sarakonradi.

**Comment by sarakonradi** [comment]
Added the information on Boost, as suggested by @terhorstd.  

> This PR does a lot more changes than adding information about nest_vars.sh, it basically reworks the Linux install instructions.

This is a classic instance of a PR evolving (more than it should) over time. As discussed in the doc meeting, it is best to keep future PRs small! Should we rename it or keep the title?

**Comment by clinssen** [comment]
Would propose to update the title and merge this so we can move on.

**Comment by sarakonradi** [comment]
> Would propose to update the title and merge this so we can move on.

Done.

**Comment by jougs** [comment]
I totally agree that we should merge this one now and suggest to mark up the remarks in a next step, which should probably also check if they are still valid, some of them seem quite outdated...

 @terhorstd: can you have a final look and push the button if you're happy?
